### PR TITLE
Add Policy API's

### PR DESCRIPTION
**Pull request by neosab**
* Adds Policy API's `GET, POST, DELETE`
* Adds a simple casbin storage adapter for loading policy entities
* Adds controller and policy entity handlers
* Adds support for `--enable-bootstrap-mode` for identity-manager (to add initial policies)
* Ability to specify a bootstrap user in install config.yaml
* some other necessary ingress changes

---
This change is [Reviewable](https://reviewable.io/reviews/vmware/dispatch/246)

### CODE REVIEW

File: charts/dispatch/charts/identity-manager/templates/deployment.yaml [hunk 1]
```diff
@@ -28,12 +28,19 @@ spec:
           imagePullPolicy: {{ default .Values.global.pullPolicy .Values.image.pullPolicy }}
           args:
             - "--cookie-name=_oauth2_proxy"
-            - "--policy-file=/etc/config/policy.csv"
             - "--host=0.0.0.0"
             - "--port=8080"
+            - "--db-file={{ default .Release.Name .Values.global.db.release }}-{{ .Values.global.db.host }}.{{ default .Release.Namespace .Values.global.db.namespace }}:{{ .Values.global.db.port }}"
```
Comment (berndtj) [inline charts/dispatch/charts/identity-manager/templates/deployment.yaml]: What do we want to make a CLI arg vs. what goes in config.json?  I want us to be consistent across services.  I just took things out of CLI args for function manager.  I don't really care one way or the other, I just want us to be consistent.

File: charts/dispatch/charts/identity-manager/templates/ingress.noauth.yaml [hunk 2]
```diff
@@ -4,23 +4,29 @@
...
   annotations:
-    {{- range $key, $value := .Values.ingress.home.annotations }}
-    {{ $key }}: {{ $value | quote }}
-    {{- end }}
     nginx.ingress.kubernetes.io/configuration-snippet: |
       error_page 403 = @403.json;
+    nginx.ingress.kubernetes.io/server-snippet: |
+      location @403.json {
+        default_type application/json;
+        return 403 '{"code": 403,"message": "Forbidden, please make sure you are logged in and have required privileges"}';
```
Comment (berndtj) [inline charts/dispatch/charts/identity-manager/templates/ingress.noauth.yaml]: 👍

File: charts/dispatch/charts/identity-manager/templates/secret.yaml [hunk 3]
```diff
@@ -9,6 +9,7 @@ metadata:
     heritage: "{{ .Release.Service }}"
 type: Opaque
 data:
+  bootstrap_user: {{ default "" .Values.bootstrapUser | trim | b64enc | quote }}
```
Comment (berndtj) [inline charts/dispatch/charts/identity-manager/templates/secret.yaml]: nit, inconsistent spacing

File: pkg/identity-manager/casbin_entity_adapter.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+func NewCasbinEntityAdapter(store entitystore.EntityStore) *CasbinEntityAdapter {
+	return &CasbinEntityAdapter{store: store}
+}
+
+func (a *CasbinEntityAdapter) LoadPolicy(model model.Model) error {
+	var policies []*Policy
+
+	// Policy status like CREATING, DELETING doesn't have any effect on
```
Comment (tenczar) [inline pkg/identity-manager/casbin_entity_adapter.go]: Doesn't have any effect on what?

File: pkg/identity-manager/policy_handler.go [hunk 5]
```diff
@@ -0,0 +1,87 @@
+///////////////////////////////////////////////////////////////////////
```
Comment (tenczar) [inline pkg/identity-manager/policy_handler.go]: This kind of confused me. The api-manager and image-manager defined this handler in the same controller.go file as the associated controller. We probably want to keep all of these consistent. I think the way you did it here makes more sense.

File: cmd/identity-manager/main.go [hunk 6]
```diff
@@ -82,9 +84,28 @@ func main() {
...
+	es, err := entitystore.NewFromBackend(
+		entitystore.BackendConfig{
+			Backend:  identitymanager.IdentityManagerFlags.DbBackend,
+			Address:  identitymanager.IdentityManagerFlags.DbFile,
+			Bucket:   identitymanager.IdentityManagerFlags.DbDatabase,
+			Username: identitymanager.IdentityManagerFlags.DbUser,
+			Password: identitymanager.IdentityManagerFlags.DbPassword,
+		})
+	if err != nil {
+		log.Fatalln(err)
```
[the remark this answers is not in the document]
Comment (kars7e) [reply cmd/identity-manager/main.go]: `log.Fatalln` runs `os.Exit(1)`, so it should bail out here

File: pkg/identity-manager/casbin_entity_adapter.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	opts := entitystore.Options{
+		Filter: entitystore.FilterExists(),
+	}
+
+	log.Debug("Reloading policies")
+	err := a.store.List(IdentityManagerFlags.OrgID, opts, &policies)
+	for _, policy := range policies {
+		// Casbin authorization rules are of the form (subject, resource, action) and hence the need to iterate over all rule fields.
+		for _, rule := range policy.Rules {
```
Comment (kars7e) [inline pkg/identity-manager/casbin_entity_adapter.go]: That's a lot of `for`s! :)

File: pkg/identity-manager/casbin_entity_adapter.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+type CasbinEntityAdapter struct {
+	store entitystore.EntityStore
+}
+
+func NewCasbinEntityAdapter(store entitystore.EntityStore) *CasbinEntityAdapter {
+	return &CasbinEntityAdapter{store: store}
+}
+
+func (a *CasbinEntityAdapter) LoadPolicy(model model.Model) error {
```
[the remark this answers is not in the document]
Comment (neosab) [reply pkg/identity-manager/casbin_entity_adapter.go]: Good point, I will alias it.

File: pkg/identity-manager/casbin_entity_adapter.go (hunk 4 above)
Comment (neosab) [reply pkg/identity-manager/casbin_entity_adapter.go]: I guess I context switched while typing that ;), I will fix it. Basically, I meant that LoadPolicy doesn't inspect the status of the policy entity. The status is more for the controllers to act upon.

File: pkg/identity-manager/casbin_entity_adapter.go (hunk 7 above)
Comment (neosab) [reply pkg/identity-manager/casbin_entity_adapter.go]: I KNEW someone would say that, anyway I will fix it with some memoization 📝

File: pkg/identity-manager/policy_handler.go (hunk 5 above)
Comment (neosab) [reply pkg/identity-manager/policy_handler.go]: Hmm, I kinda followed the suite from event manager. While working on this I noticed a lot of minor inconsistencies within different microservices. We need to fix them.

File: charts/dispatch/charts/identity-manager/templates/deployment.yaml (hunk 1 above)
Comment (neosab) [reply charts/dispatch/charts/identity-manager/templates/deployment.yaml]: I still see command args being passed to the function manager's deployment https://github.com/vmware/dispatch/blob/master/charts/dispatch/charts/function-manager/templates/deployment.yaml#L31

Did you mean something else?

File: charts/dispatch/charts/identity-manager/templates/secret.yaml (hunk 3 above)
Comment (neosab) [reply charts/dispatch/charts/identity-manager/templates/secret.yaml]: Will fix it thanks!

File: cmd/identity-manager/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tenczar) [reply cmd/identity-manager/main.go]: Not sure how I feel about that. I guess it is convenient, but the doc comment should really mention that it calls os.Exit(1).

**Pull request merged**